azurekvp/src/tracing.rs

```diff
@@ -0,0 +1,219 @@
+) -> String {
+    format!(
+        "{}|{}|{}|{}",
+        EVENT_PREFIX, event_level, event_name, span_id
+    )
+}
+
+fn encode_kvp_item(key: &str, value: &str) -> Vec<Vec<u8>> {
+    let key_bytes = key.as_bytes();
+    let value_bytes = value.as_bytes();
+
+    let key_len = key_bytes.len().min(HV_KVP_EXCHANGE_MAX_KEY_SIZE);
+    let mut key_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_KEY_SIZE];
+    key_buf[..key_len].copy_from_slice(&key_bytes[..key_len]);
+
+    if value_bytes.len() <= HV_KVP_AZURE_MAX_VALUE_SIZE {
+        let mut value_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_VALUE_SIZE];
+        let value_len = value_bytes.len().min(HV_KVP_EXCHANGE_MAX_VALUE_SIZE);
+        value_buf[..value_len].copy_from_slice(&value_bytes[..value_len]);
+
+        vec![encode_kvp_slice(key_buf, value_buf)]
+    } else {
+        println!("Value exceeds max size, splitting into multiple slices.");
+
+        let mut kvp_slices = Vec::new();
+        let mut start = 0;
+        while start < value_bytes.len() {
+            let end =
+                (start + HV_KVP_AZURE_MAX_VALUE_SIZE).min(value_bytes.len());
+            let mut value_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_VALUE_SIZE];
+            value_buf[..end - start].copy_from_slice(&value_bytes[start..end]);
+
+            kvp_slices.push(encode_kvp_slice(key_buf.clone(), value_buf));
+            start += HV_KVP_AZURE_MAX_VALUE_SIZE;
+        }
+        kvp_slices
+    }
+}
+
+fn encode_kvp_slice(key: Vec<u8>, value: Vec<u8>) -> Vec<u8> {
+    let mut buffer = Vec::with_capacity(
+        HV_KVP_EXCHANGE_MAX_KEY_SIZE + HV_KVP_EXCHANGE_MAX_VALUE_SIZE,
+    );
+    buffer.extend_from_slice(&key);
+    buffer.extend_from_slice(&value);
+    buffer
+}
+
+pub fn decode_kvp_item(
+    record_data: &[u8],
+) -> Result<(String, String), &'static str> {
+    let record_data_len = record_data.len();
+    let expected_len =
+        HV_KVP_EXCHANGE_MAX_KEY_SIZE + HV_KVP_EXCHANGE_MAX_VALUE_SIZE;
+
+    if record_data_len != expected_len {
+        return Err("record_data len not correct.");
+    }
+
+    let key = String::from_utf8(
+        record_data[0..HV_KVP_EXCHANGE_MAX_KEY_SIZE].to_vec(),
+    )
+    .unwrap_or_else(|_| String::new())
+    .trim_end_matches('\x00')
+    .to_string();
+
+    let value = String::from_utf8(
+        record_data[HV_KVP_EXCHANGE_MAX_KEY_SIZE..record_data_len].to_vec(),
+    )
+    .unwrap_or_else(|_| String::new())
+    .trim_end_matches('\x00')
+    .to_string();
+
+    Ok((key, value))
+}
+
+pub fn truncate_guest_pool_file(kvp_file: &Path) -> Result<(), Error> {
+    let boot_time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .map_err(|e| Error::new(std::io::ErrorKind::Other, e))?
+        .as_secs()
+        - get_uptime()?.as_secs();
+
+    match kvp_file.metadata() {
+        Ok(metadata) => {
+            if metadata.mtime() < boot_time as i64 {
+                OpenOptions::new()
+                    .write(true)
+                    .truncate(true)
```

Comment:
I know this is how cloud-init does it, but I don't think we should be in the business of truncating the KVP pool.  If the pool shouldn't persist across boots, it needs to be moved to /run or similar.

But then the question becomes can we handle a persistent KVP pool rather than assuming clean slate...

Not asking for a change, but something to ponder (how would we support a different path if we fixed this).